src/framework/components/collision/system.js

```diff
@@ -355,27 +357,33 @@ class CollisionMeshSystemImpl extends CollisionSystemImpl {
             const numTriangles = mesh.primitive[0].count / 3;
 
             const v1 = new Ammo.btVector3();
-            const v2 = new Ammo.btVector3();
-            const v3 = new Ammo.btVector3();
             let i1, i2, i3;
 
             const base = mesh.primitive[0].base;
             triMesh = new Ammo.btTriangleMesh();
             this.system._triMeshCache[mesh.id] = triMesh;
 
-            for (let i = 0; i < numTriangles; i++) {
-                i1 = indices[base + i * 3] * stride;
-                i2 = indices[base + i * 3 + 1] * stride;
-                i3 = indices[base + i * 3 + 2] * stride;
-                v1.setValue(positions[i1], positions[i1 + 1], positions[i1 + 2]);
-                v2.setValue(positions[i2], positions[i2 + 1], positions[i2 + 2]);
-                v3.setValue(positions[i3], positions[i3 + 1], positions[i3 + 2]);
-                triMesh.addTriangle(v1, v2, v3, true);
+            const numVertices = positions.length / stride;
+
+            const indexedArray = triMesh.getIndexedMeshArray();
+            indexedArray.at(0).m_numTriangles = numTriangles;
+
+            for (let i = 0; i < numVertices; i++) {
+                v1.setValue(positions[i * stride], positions[i * stride + 1], positions[i * stride + 2]);
+                triMesh.findOrAddVertex(v1, checkDuplicates);
```

Comment:
based on this (not sure this is the latest version, not an expert here) it seems `findOrAddVertex` and `addIndex` are internal, and so perhaps we should just keep using addTriangle and pass removeDuplicateVertices as false to it.
https://github.com/bulletphysics/bullet3/blob/master/src/BulletCollision/CollisionShapes/btTriangleMesh.h#L63-L66

One additional reason I suggest this is that I'm not sure the way those functions are currently used is correct. `findOrAddVertex` returns a new index in case a duplicate was found, but we're not handling it, which means that this code will access vertices that do not exist (our of range).
https://github.com/bulletphysics/bullet3/blob/39b8de74df93721add193e5b3d9ebee579faebf8/src/BulletCollision/CollisionShapes/btTriangleMesh.cpp#L86

Reply:
Hmm, so we have 2 methods here:
1. using the previous `addTriangle` method with disabled duplicates check
2. using `findOrAddVertex` and `addIndex` with disabled duplicates check

The second one is about 10% faster than the first one (which can be around 5-15ms for complex meshes). Considering there is a faster method, do we still want to use the `addTriangle`? I'm ok with both options. Both of them seem to be order of magnitudes faster than the current method with forced duplicates check.

As for the returned index in case of a duplicate found - I am not sure if I understand. By new index do you mean an old index that was incremented? If that would be the case, then I'd agree - we could potentially point out of bounds. However, I don't see it being the case. This line returns an index of an existing vertex when a duplicate is found, not a new one:
https://github.com/bulletphysics/bullet3/blob/39b8de74df93721add193e5b3d9ebee579faebf8/src/BulletCollision/CollisionShapes/btTriangleMesh.cpp#L115

However, said that, I do think there should be a guard. If we consider worse case scenario - a user provides a mesh with many duplicate vertices and manually disables the duplicate test. This will generate a collision mesh, but with triangles that would not match the shape. If we do use the second method, then to mitigate this case, we could use the first method for the default `checkVertexDuplicates` = `true`, and go for the second one, when it is explicitly disabled (i.e. the developer knows the mesh is clean).